src/Simulation/qdk_sim_rs/src/linalg/expm.rs

```diff
@@ -0,0 +1,212 @@
+    E: std::convert::From<QdkSimError>,
+{
+    // D_pq(A) ≔ Σ_j=0^q [(p + q - j)! p!] / [(p + q)! j! (q - j)!] (-A)^j
+    let mut result = mtx.eye_like().to_owned();
+    for j in 0..q + 1 {
+        let arr: [i64; 0] = [];
+        assert!(arr.len() == 0);
+        let coeff = (approximate_factorial::<_, f64>(p + q - j)
+            * approximate_factorial::<_, f64>(p))
+            / (approximate_factorial::<_, f64>(p + q)
+                * approximate_factorial::<_, f64>(j)
+                * approximate_factorial::<_, f64>(q - j));
+        let coeff = A::from(coeff).unwrap();
+        let mtx_pow = (-mtx.to_owned()).matrix_power(j.into())?;
+        let term = mtx_pow * coeff;
+        result = result + term;
+    }
+    Ok(result)
+}
+
+#[cfg(feature = "pade")]
+fn pade_n<A: Scalar + ScalarOperand, S: Data + RawData<Elem = A>, E>(
+    mtx: &ArrayBase<S, Ix2>,
+    p: u32,
+    q: u32,
+) -> Result<ArrayBase<OwnedRepr<A>, Ix2>, E>
+where
+    ArrayBase<S, Ix2>: Inv<Error = E> + Identity,
+    E: std::convert::From<QdkSimError>,
+{
+    // N_pq(A) ≔ Σ_j=0^p [(p + q - j)! p!] / [(p + q)! j! (p - j)!] A^j
+    // [(p + q - 0)! p!] / [(p + q)! (p)!]
+    let mut result = mtx.eye_like().to_owned();
+    for j in 0..p + 1 {
+        let coeff = (approximate_factorial::<_, f64>(p + q - j)
+            * approximate_factorial::<_, f64>(p))
+            / (approximate_factorial::<_, f64>(p + q)
+                * approximate_factorial::<_, f64>(j)
+                * approximate_factorial::<_, f64>(p - j));
+        let coeff = A::from(coeff).unwrap();
+        let mtx_pow = mtx.to_owned().matrix_power(j.into())?;
+        let term = mtx_pow * coeff;
+        result = result + term;
+    }
+    Ok(result)
+}
+
+#[cfg(test)]
+mod tests {
+    use approx::assert_abs_diff_eq;
+    use cauchy::c64;
+    use ndarray::array;
+    use num_traits::{One, Zero};
+
+    use crate::{
+        c64,
+        error::QdkSimError,
+        linalg::{decompositions::ExplicitEigenvalueDecomposition, Expm},
+    };
+
+    #[cfg(feature = "pade")]
+    #[test]
+    fn expm_works_for_rz() -> Result<(), QdkSimError> {
+        let hz =
+            c64::new(0.0, -1.0) * array![[c64::one(), c64::zero()], [c64::zero(), -c64::one()]];
+        println!("hz = {hz}");
+        let u = hz.expm();
+        // FIXME: returns the wrong answer, so we print while we diagnose.
+        println!("{:?}", u);
+        panic!("Test currently fails.");
+    }
+
+    #[test]
+    fn explicit_eig_works_for_ry() -> Result<(), QdkSimError> {
+        let hy = ExplicitEigenvalueDecomposition {
+            values: array![c64!(1.0), c64!(-1.0)],
+            vectors: array![
+                [
+                    c64::new(-0.0, -0.7071067811865474),
+                    c64::new(0.7071067811865475, 0.0)
+                ],
+                [
+                    c64::new(0.7071067811865476, 0.0),
+                    c64::new(0.0, -0.7071067811865475)
+                ],
+            ],
```

Comment:
Fragment "0.7071067811865475" occurs more than once. Consider de-duplicating. Applicable to more than one file.
Consider explaining the values for the "magic numbers" (in the code comments), such that the reader can make sure that the value is correct.

Reply:
Good point, thanks; these were generated by using Python to produce unit test cases, but could now be replaced with [`std::f64::consts::FRAC_1_SQRT_2`](https://doc.rust-lang.org/std/f64/consts/constant.FRAC_1_SQRT_2.html).

(As a side note, 0.707 comes up a _lot_ in quantum applications, such that I'm slightly biased — I no longer really tend to see it as a magic number.)